hanish/bot.py

```diff
@@ -0,0 +1,409 @@
+        self.shutdown = True
+
+    def read_rtm_channel(self):
+        """
+        Reads the RTM channel, parses output, filters messages based on type
+        and if they are directed at the bot. If so, passes those messages on
+        to the message handler command.
+        """
+
+        # Check for messages on the wire
+        # TODO: as soon as we connect, the last message comes through.
+        messages = self.slack.rtm_read()
+        if messages and len(messages) > 0:
+            # If there are messages, begin filtering and parsing.
+            for msg in messages:
+                # Filter only messages from other users
+                if msg["type"] == "message":
+                    # Check if an @botid appears in the message
+                    if self.atbotid.search(msg.get("text", "")):
+                        # Handle the atbotid message
+                        self.handle_message(msg)
+
+    def handle_message(self, msg):
+        """
+        Handles messages that are directed at the bot by parsing commands.
+        """
+        # Log the message received
+        # TODO: change to standardized logging functionality
+        print("[{}] {}".format(msg["ts"], msg["text"]))
+
+        # Track if an unknown command has occurred
+        unknown = True
+
+        # Determine if the message matches any commands
+        for name, command in self.commands.items():
+            match = command.search(msg["text"])
+            if match is not None:
+                unknown = False # We've found at least one command match.
+
+                # Get the method that handles this command
+                handler = getattr(self, "handle_{}_command".format(name), None)
+                if handler is None:
+                    raise NotImplementedError(
+                        "{} command not yet handled".format(name)
+                    )
+
+                try:
+                    # Handle the command
+                    handler(msg)
+                except HanishException as e:
+                    # Respond with hanish exceptions
+                    response = "Sorry, there was a problem with your request: {}".format(e)
+                    self.post(msg['channel'], response)
+                except Exception as e:
+                    # Fatal exception has occurred
+                    self.post(msg['channel'], "A fatal exception has occurred!")
+
+        # Handle any unknown commands
+        if unknown:
+            self.handle_unknown_command(msg)
+
+    def handle_weather_command(self, msg):
+        """
+        Have the chatbot respond to the weather command.
+        """
+        # Parse the weather command arguments
+        # NOTE: only the first weather command will be matched
+        # TODO: handle all unique weather commands/arguments
+        match  = self.commands['weather'].search(msg['text'])
+        period = match.group(1).lower()
+        weather = self.weather()
+
+        if period == "now":
```

Comment:
Yeh, that was the original goal, but I realized that if you write `weather now weather now` then you would get two responses, and that wasn't ideal. My preference is a jump table or the dict of lambdas instead of if/else blocks, if that helps!